File: Leetcode/Max Points on a Line.cpp
/**
 * Definition for a point.
 * struct Point {
 *     int x;
 *     int y;
 *     Point() : x(0), y(0) {}
 *     Point(int a, int b) : x(a), y(b) {}
 * };
 */
class Solution {
public:
    const double INF = 1e60;
    int maxPoints(vector<Point> &points) {
        map<double,int> mp;
        
        int n = points.size();
        int same,ans = 0;
        for (int i = 0; i < n; ++i) {
            same = 0;
            mp.clear();
            mp[INF] = 0;
            for (int j = 0; j < n; ++j) {
                if (points[i].x == points[j].x && points[i].y == points[j].y) {
                    ++same;
                    continue;
                }
                if (points[i].x == points[j].x) {
                    mp[INF]++;
                } else  {
                    double tmp = (double)(points[i].y-points[j].y)/(double)(points[i].x-points[j].x);
                    mp[tmp]++;
                }
            }
            for (auto it = mp.begin(); it != mp.end(); ++it) {
                ans = max(ans,it->second + same);
            }
        }
        return ans;
    }
};

File: Leetcode/Max Points on a Line.py
# Definition for a point
# class Point:
#     def __init__(self, a=0, b=0):
#         self.x = a
#         self.y = b

class Solution:
    # @param points, a list of Points
    # @return an integer
    def maxPoints(self, points):
        n = len(points)
        if n < 3: return n
        ret = 0
        for i in xrange(n):
            slope = {'inf': 0}
            samePoint = 0
            for j in xrange(n):
                if points[i].x == points[j].x and points[i].y == points[j].y:
                    samePoint += 1
                elif points[i].x == points[j].x:
                    slope['inf'] += 1
                else:
                     k = 1.0 * (points[i].y - points[j].y) / (points[i].x - points[j].x)
                     slope[k] = 1 if k not in slope else slope[k] + 1
            ret = max(ret, max(slope.values()) + samePoint)
        return ret